### ISSUE TITLE
DataStore with @auth - Sync error subscription failed ... Missing field argument owner

### ISSUE THREAD

**Issue by chuckatc** [aws-amplify/amplify-cli#3048]
***tl;dr*** Is it necessary to configure DataStore to include `owner` for syncing when using owner auth? If so, what's the syntax or can anyone point to a working example?

**Which Category is your question related to?**

Amplify CLI, DataStore, auth, GraphQL API

**Amplify CLI Version** 

4.6.0

**What AWS Services are you utilizing?**

Amplify CLI, Datastore, AppSync, Cognito User Pools

**Provide additional details e.g. code snippets**

Hi all, attempting to add simple owner authorization to this DataStore demo https://github.com/sebsto/amplify-datastore-js-e2e . 

***Changes to the demo repo***

`@auth` directive was added to schema:

```graphql
enum PostStatus {
  ACTIVE
  INACTIVE
}

type Post @model @auth(rules: [{ allow: owner }]) {
  id: ID!
  title: String!
  rating: Int!
  status: PostStatus!
}
```

App.js updated for auth:

```javascript
export default withAuthenticator(App, true);
```

The usual `amplify init`, `amplify add auth` (Cognito User Pools), `npm run amplify-modelgen`, `npm run amplify-push`, possibly other commands in there

Signing up and signing in...

***Result***

Browser fails to sync records giving the following in the console:

```
[WARN] 21:10.98 DataStore - Sync error subscription failed Connection failed: {"errors":[{"message":"Validation error of type MissingFieldArgument: Missing field argument owner @ 'onCreatePost'"}]}
```

which is followed after ~20 seconds by:

```
AWSAppSyncRealTimeProvider.ts:506 Uncaught TypeError: Cannot read property 'observer' of undefined
    at AWSAppSyncRealTimeProvider._timeoutStartSubscriptionAck (AWSAppSyncRealTimeProvider.ts:506)
    at AWSAppSyncRealTimeProvider.ts:318
```

***Please note***
* same errors when commenting out the useEffect containing the `DataStore.observe(Post).subscribe(...)` call and clicking buttons to trigger a `DataStore.query()` instead, so it seems DataStore is trying to set up subscriptions itself
* same errors when adding `owner` to the call to `DataStore.save()` apart from records created in IndexedDB now showing `owner`
* from the AppSync console, querying and mutating Posts as the authenticated user works fine and includes an `owner` field

So, is it necessary to configure DataStore explicitly to use `owner`? e.g. an equivalent to `API.graphql(graphqlOperation(onUpdateNote, { owner })).subscribe(...)` in the GraphQL API? Or should DataStore pick that up automagically and the issue is somewhere else?

(Also wondering, are there any samples/demos/docs using DataStore together with auth yet? I understand it's early days for DataStore but haven't found any so far)

**Comment by kaustavghosh06** [comment]
@chuckatc We're actively working on this and expect a fix for this very soon.
cc @yuth  @elorzafe

**Comment by dconlon-monigo** [comment]
I'm hitting this too, or at least something similar, take the following slightly extended example,

```
Type Task
    @model
    @auth(rules: [
        {allow: groups, groups: ["SuperUser"], queries: [get, list], mutations: [create, update, delete]},
        {allow: owner, queries: [get, list], mutations: [create]}
      ])
 {
    id: ID!
    title: String!
    description: String
    status: String
 }
```

I'd expect an owner to be able to subscribe to onCreateTask, and get only their 'Tasks', and a user in the SuperUser group to be able to subscribe to **_all_** onCreateTask events from all users. As it is today the owner is a required argument to the generated subscription, so when a member of the SuperUser group subscribes with their user as owner they only see their create events not all create events.

**Comment by dbhagen** [comment]
I think I'm hitting something similar to my app. Setup is pretty much the same as the original author, Javascript client, amplify generated backend and models.

Schema is this:
`type Timeline
  @model
  @searchable
  @auth(
    rules: [
      { allow: groups, groups: ["SuperUser"] }
      { allow: owner }
    ]
  ) {
  id: ID!
  name: String!
  startDate: AWSDate!
  endDate: AWSDate!
  investment401k: [Investment401k]
    @connection(name: "TimelineInvestment401ks")
  income: [Income] @connection(name: "TimelineIncomes")
}
type Income
  @model
  @searchable
  @auth(
    rules: [
      { allow: groups, groups: ["SuperUser"] }
      { allow: owner }
    ]
  ) {
  id: ID!
  startDate: AWSDate!
  endDate: AWSDate
  rate: Float!
  object: AWSJSON
  timeline: Timeline! @connection(name: "TimelineIncomes")
}
type Investment401k
  @model
  @searchable
  @auth(
    rules: [
      { allow: groups, groups: ["SuperUser"] }
      { allow: owner }
    ]
  ) {
  id: ID!
  object: AWSJSON
  timeline: Timeline! @connection(name: "TimelineInvestment401ks")
}
type UserData
  @model
  @searchable
  @auth(
    rules: [
      { allow: groups, groups: ["SuperUser"] }
      { allow: owner }
    ]
  ) {
  id: ID!
  birthdate: AWSDate!
  retirementdate: AWSDate!
  profileImageUrl: String
  profileImageUncroppedUrl: String
  profileImageManipulations: AWSJSON
}`

aws-exports.js, which is generated by Amplify, is this (test environment, nothing production):
`// WARNING: DO NOT EDIT. This file is automatically generated by AWS Amplify. It will be overwritten.

const awsmobile = {
    "aws_project_region": "us-east-1",
    "aws_cognito_identity_pool_id": "us-east-1:85e7460b-feab-4f21-87b7-a9a265125c77",
    "aws_cognito_region": "us-east-1",
    "aws_user_pools_id": "us-east-1_txxH6nvSj",
    "aws_user_pools_web_client_id": "7t9mnj8a2pqfrju2e0tjh40ka4",
    "oauth": {
        "domain": "modelmymoney655c1605-655c1605-localdev.auth.us-east-1.amazoncognito.com",
        "scope": [
            "phone",
            "email",
            "openid",
            "profile",
            "aws.cognito.signin.user.admin"
        ],
        "redirectSignIn": "http://localhost/auth/signin/",
        "redirectSignOut": "http://localhost/auth/signout/",
        "responseType": "code"
    },
    "federationTarget": "COGNITO_USER_POOLS",
    "aws_appsync_graphqlEndpoint": "https://aitvi3puvfcwfkhqmhmsiw6i3q.appsync-api.us-east-1.amazonaws.com/graphql",
    "aws_appsync_region": "us-east-1",
    "aws_appsync_authenticationType": "AMAZON_COGNITO_USER_POOLS",
    "aws_appsync_apiKey": "da2-g47wpfle4rhzrotlqnfu6cgew4",
    "aws_user_files_s3_bucket": "model-my-money947365a9c1734517af86f8ba26db26a7money-localdev",
    "aws_user_files_s3_bucket_region": "us-east-1"
};


export default awsmobile;`

Relevant lines from main.js
`import Amplify from 'aws-amplify';
import awsconfig from '@/aws-exports';
Amplify.configure(awsconfig);`

And the generated schema.js in the /models/ folder:
`export const schema = {
    "models": {
        "Timeline": {
            "syncable": true,
            "name": "Timeline",
            "pluralName": "Timelines",
            "attributes": [
                {
                    "type": "model",
                    "properties": {}
                },
                {
                    "type": "searchable",
                    "properties": {}
                },
                {
                    "type": "auth",
                    "properties": {
                        "rules": [
                            {
                                "groupClaim": "cognito:groups",
                                "provider": "userPools",
                                "allow": "groups",
                                "groups": [
                                    "SuperUser"
                                ],
                                "operations": [
                                    "create",
                                    "update",
                                    "delete"
                                ]
                            },
                            {
                                "provider": "userPools",
                                "ownerField": "owner",
                                "allow": "owner",
                                "identityClaim": "cognito:username",
                                "operations": [
                                    "create",
                                    "update",
                                    "delete"
                                ]
                            }
                        ]
                    }
                }
            ],
            "fields": {
                "id": {
                    "name": "id",
                    "isArray": false,
                    "type": "ID",
                    "isRequired": true,
                    "attributes": []
                },
                "name": {
                    "name": "name",
                    "isArray": false,
                    "type": "string",
                    "isRequired": true,
                    "attributes": []
                },
                "startDate": {
                    "name": "startDate",
                    "isArray": false,
                    "type": "AWSDate",
                    "isRequired": true,
                    "attributes": []
                },
                "endDate": {
                    "name": "endDate",
                    "isArray": false,
                    "type": "AWSDate",
                    "isRequired": true,
                    "attributes": []
                },
                "investment401k": {
                    "name": "investment401k",
                    "isArray": true,
                    "type": {
                        "model": "Investment401k"
                    },
                    "isRequired": false,
                    "attributes": [],
                    "association": {
                        "connectionType": "HAS_MANY",
                        "associatedWith": "timeline"
                    }
                },
                "income": {
                    "name": "income",
                    "isArray": true,
                    "type": {
                        "model": "Income"
                    },
                    "isRequired": false,
                    "attributes": [],
                    "association": {
                        "connectionType": "HAS_MANY",
                        "associatedWith": "timeline"
                    }
                }
            }
        },
        "Investment401k": {
            "syncable": true,
            "name": "Investment401k",
            "pluralName": "Investment401ks",
            "attributes": [
                {
                    "type": "model",
                    "properties": {}
                },
                {
                    "type": "searchable",
                    "properties": {}
                },
                {
                    "type": "auth",
                    "properties": {
                        "rules": [
                            {
                                "groupClaim": "cognito:groups",
                                "provider": "userPools",
                                "allow": "groups",
                                "groups": [
                                    "SuperUser"
                                ],
                                "operations": [
                                    "create",
                                    "update",
                                    "delete"
                                ]
                            },
                            {
                                "provider": "userPools",
                                "ownerField": "owner",
                                "allow": "owner",
                                "identityClaim": "cognito:username",
                                "operations": [
                                    "create",
                                    "update",
                                    "delete"
                                ]
                            }
                        ]
                    }
                }
            ],
            "fields": {
                "id": {
                    "name": "id",
                    "isArray": false,
                    "type": "ID",
                    "isRequired": true,
                    "attributes": []
                },
                "object": {
                    "name": "object",
                    "isArray": false,
                    "type": "AWSJSON",
                    "isRequired": false,
                    "attributes": []
                },
                "timeline": {
                    "name": "timeline",
                    "isArray": false,
                    "type": {
                        "model": "Timeline"
                    },
                    "isRequired": true,
                    "attributes": [],
                    "association": {
                        "connectionType": "BELONGS_TO",
                        "targetName": "investment401kTimelineId"
                    }
                }
            }
        },
        "Income": {
            "syncable": true,
            "name": "Income",
            "pluralName": "Incomes",
            "attributes": [
                {
                    "type": "model",
                    "properties": {}
                },
                {
                    "type": "searchable",
                    "properties": {}
                },
                {
                    "type": "auth",
                    "properties": {
                        "rules": [
                            {
                                "groupClaim": "cognito:groups",
                                "provider": "userPools",
                                "allow": "groups",
                                "groups": [
                                    "SuperUser"
                                ],
                                "operations": [
                                    "create",
                                    "update",
                                    "delete"
                                ]
                            },
                            {
                                "provider": "userPools",
                                "ownerField": "owner",
                                "allow": "owner",
                                "identityClaim": "cognito:username",
                                "operations": [
                                    "create",
                                    "update",
                                    "delete"
                                ]
                            }
                        ]
                    }
                }
            ],
            "fields": {
                "id": {
                    "name": "id",
                    "isArray": false,
                    "type": "ID",
                    "isRequired": true,
                    "attributes": []
                },
                "startDate": {
                    "name": "startDate",
                    "isArray": false,
                    "type": "AWSDate",
                    "isRequired": true,
                    "attributes": []
                },
                "endDate": {
                    "name": "endDate",
                    "isArray": false,
                    "type": "AWSDate",
                    "isRequired": false,
                    "attributes": []
                },
                "rate": {
                    "name": "rate",
                    "isArray": false,
                    "type": "Float",
                    "isRequired": true,
                    "attributes": []
                },
                "object": {
                    "name": "object",
                    "isArray": false,
                    "type": "AWSJSON",
                    "isRequired": false,
                    "attributes": []
                },
                "timeline": {
                    "name": "timeline",
                    "isArray": false,
                    "type": {
                        "model": "Timeline"
                    },
                    "isRequired": true,
                    "attributes": [],
                    "association": {
                        "connectionType": "BELONGS_TO",
                        "targetName": "incomeTimelineId"
                    }
                }
            }
        },
        "UserData": {
            "syncable": true,
            "name": "UserData",
            "pluralName": "UserData",
            "attributes": [
                {
                    "type": "model",
                    "properties": {}
                },
                {
                    "type": "searchable",
                    "properties": {}
                },
                {
                    "type": "auth",
                    "properties": {
                        "rules": [
                            {
                                "groupClaim": "cognito:groups",
                                "provider": "userPools",
                                "allow": "groups",
                                "groups": [
                                    "SuperUser"
                                ],
                                "operations": [
                                    "create",
                                    "update",
                                    "delete"
                                ]
                            },
                            {
                                "provider": "userPools",
                                "ownerField": "owner",
                                "allow": "owner",
                                "identityClaim": "cognito:username",
                                "operations": [
                                    "create",
                                    "update",
                                    "delete"
                                ]
                            }
                        ]
                    }
                }
            ],
            "fields": {
                "id": {
                    "name": "id",
                    "isArray": false,
                    "type": "ID",
                    "isRequired": true,
                    "attributes": []
                },
                "birthdate": {
                    "name": "birthdate",
                    "isArray": false,
                    "type": "AWSDate",
                    "isRequired": true,
                    "attributes": []
                },
                "retirementdate": {
                    "name": "retirementdate",
                    "isArray": false,
                    "type": "AWSDate",
                    "isRequired": true,
                    "attributes": []
                },
                "profileImageUrl": {
                    "name": "profileImageUrl",
                    "isArray": false,
                    "type": "string",
                    "isRequired": false,
                    "attributes": []
                },
                "profileImageUncroppedUrl": {
                    "name": "profileImageUncroppedUrl",
                    "isArray": false,
                    "type": "string",
                    "isRequired": false,
                    "attributes": []
                },
                "profileImageManipulations": {
                    "name": "profileImageManipulations",
                    "isArray": false,
                    "type": "AWSJSON",
                    "isRequired": false,
                    "attributes": []
                }
            }
        }
    },
    "enums": {},
    "version": "c3d791373597b9b143173b56a57ae474"
};`

**Comment by CodySwannGT** [comment]
Not much to add - same as the others. Fix or a workaround would be fantastic as we're in production with no path forward.

**Comment by elorzafe** [comment]
@dbhagen @chuckatc @dconlon-monigo @CodySwannGT the fix for the library is already published on @latest tag on npm.

**Comment by ashteya** [comment]
@elorzaf I was getting the same problem as @chuckatc, just updated to `@aws-amplify/datastore` version 1.04 and now the sync is not working at all anymore.

I set log level to DEBUG and on 1.0.3. I can see it's doing the sync, but on 1.0.4. there is nothing logged for it.

Do we need to configure the sync?

**Comment by mkaschke** [comment]
same here!

**Comment by elorzafe** [comment]
@ashteya @mkaschke can you share your package.json and schema.graphql files?

**Comment by mkaschke** [comment]
@elorzafe sure!

package.json

```
{
  "name": "cognito-amplify-custom-auth",
  "version": "0.0.1",
  "description": "TBD",
  "main": "index.js",
  "repository": "",
  "private": true,
  "author": "mkaschke",
  "license": "ISC",
  "scripts": {
    "postinstall": "cp -R node_modules/semantic-ui-css/semantic.min.css node_modules/semantic-ui-css/themes style",
    "start": "node ./node_modules/webpack-dev-server/bin/webpack-dev-server.js",
    "test": "mocha --compilers js:babel-core/register --require ./test/test_helper.js --recursive ./test",
    "test:watch": "npm run test -- --watch",
    "build": "webpack --progress -p"
  },
  "devDependencies": {
    "@babel/core": "^7.8.0",
    "@babel/plugin-proposal-class-properties": "^7.8.0",
    "@babel/plugin-transform-runtime": "^7.8.0",
    "@babel/preset-env": "^7.8.2",
    "@babel/preset-react": "^7.8.0",
    "babel-eslint": "^10.0.1",
    "babel-loader": "^8.0.4",
    "chai": "^4.2.0",
    "css-loader": "^3.4.2",
    "eslint": "^5.9.0",
    "eslint-config-standard": "^12.0.0",
    "eslint-config-standard-react": "^7.0.2",
    "eslint-plugin-import": "^2.20.0",
    "eslint-plugin-node": "^8.0.0",
    "eslint-plugin-promise": "^4.0.1",
    "eslint-plugin-react": "^7.17.0",
    "eslint-plugin-standard": "^4.0.0",
    "jsdom": "^13.0.0",
    "json-loader": "^0.5.7",
    "mocha": "^5.2.0",
    "react-dates": "^21.5.1",
    "webpack": "^4.41.5",
    "webpack-cli": "^3.1.2",
    "webpack-dev-server": "^3.10.1"
  },
  "dependencies": {
    "amplify-cli": "^1.0.0",
    "aws-amplify": "^1.3.3",
    "aws-amplify-react": "^2.6.3",
    "aws-sdk": "^2.601.0",
    "babel-polyfill": "^6.26.0",
    "history": "^4.10.1",
    "moment": "^2.24.0",
    "prop-types": "^15.7.2",
    "react": "^16.12.0",
    "react-addons-test-utils": "^15.6.2",
    "react-dom": "^16.12.0",
    "react-redux": "^7.1.3",
    "react-router": "^5.0.0",
    "react-router-dom": "^5.0.0",
    "redux": "^4.0.5",
    "redux-form": "^8.2.3",
    "redux-promise": "^0.6.0",
    "redux-thunk": "^2.3.0",
    "semantic-ui-css": "^2.4.1",
    "semantic-ui-react": "^0.86.0",
    "style-loader": "^0.23.1"
  }
}

```

schema.graphql

```
type vehicle @auth(rules: [{ allow: owner, operations: [read, update] }]) @model @searchable  {
  id: ID!
  location: String
  city: String
  comment: String
  startDate: String
  endDate: String
  creator: String
}

type Subscription @auth(rules: [{allow: owner}]) @model @searchable  {
  id: ID!
  location: String
  city: String
  status: String
}

```

**Comment by elorzafe** [comment]
@mkaschke can you rename the type Subscription from annotated schema. That is overwriting the Subscription generated by CLI

**Comment by elorzafe** [comment]
@mkaschke you should also add `@aws-amplify/datastore` to your `package.json` file.

**Comment by chuckatc** [comment]
@elorzafe thanks very much for the attention on this. 

I'll be able to retry the modified demo with @latest later next week (currently on the road).

**Comment by yuth** [comment]
@chuckatc @mkaschke Please feel free to re-open/comment in this thread if you still facing issues

**Comment by alex-vladut** [comment]
I'm still getting an error if I use a custom `ownerField` in `@auth` decorator. Basically, with the following GraphQL schema it all works fine - default `owner` field:
```
type User @model @auth(rules: [{ allow: owner }]) {
  id: ID!
  name: String!
  email: AWSEmail!
  owner: String # required for subscriptions
}
```
but when I set the `ownerField` to be the `id` field:
```
type User @model @auth(rules: [{ allow: owner, ownerField: "id" }]) {
  id: ID!
  name: String!
  email: AWSEmail!
}
```
it throws the following exception while DataStore attempts to sync local data to the cloud:
```
{
   "errors":[
      {
         "path":["createUser"],
         "data":null,
         "errorType":"Unauthorized",
         ...
         "message":"Not Authorized to access createUser on type User"
      }
   ]
}
```
From what I can tell, DataStore assigns a random ID to the local model created (not the owner ID as it's supposed to), so when trying to sync it with AppSync will get an error due to the constraint on ownerField.
This is how I use DataStore to create a new user:
```
const init = async () => {
  const authenticatedUser = await Auth.currentAuthenticatedUser();
  const user = await DataStore.query(User, authenticatedUser.username);
  if (!user) {
    const { name, email } = authenticatedUser.attributes;
    await DataStore.save(new User({ name, email }));
  }
};
```
and this is the version of Amplify DataStore I am using in package.json:
```
"@aws-amplify/datastore": "^1.0.4",
```

**Comment by elorzafe** [comment]
@alex-vladut you are correct, DataStore generate random ids internally. 

Using `id` as `ownerField` is currently not supported. What is the use case you are trying to accomplish?

**Comment by alex-vladut** [comment]
Thanks for reply @elorzafe. At the moment I am mostly exploring DataStore in order to understand how it works and if I'll be able to use it in a prod environment. Here I try to implement a simple chat app making use of DataStore, and the specific use case listed there is to create a new User item  whenever a user signs up for the first time. The item/entity is required to store some user specific settings as well as to link a user to their chat rooms and messages.
Basically I wanted to follow the tutorial provided by Nader Dabit and port it over to Amplify DataStore:
https://dev.to/dabit3/building-chatt---a-real-time-multi-user-graphql-chat-app-3jik

**Comment by chuckatc** [comment]
@elorzafe , pardon the wait, was stuck on #3300 . Now unblocked, so will test original project soonishly.

fwiw, this [recent DataStore demo](https://medium.com/open-graphql/create-a-multiuser-graphql-crud-l-app-in-5-minutes-with-the-amplify-datastore-902764f27404) from Ed Lima uses `@auth(rules: [{ allow: owner }])` the same way so will try that as well.

**Comment by yuth** [comment]
Please feel free to comment in the thread if auth does not work with DataStore

**Comment by richardfan1126** [comment]
@alex-vladut is correct. You save my life!!

**Comment by tomheno** [comment]
Hi ! I've been struggling with this issue for a while
### Using
`"@aws-amplify/datastore": "^2.0.8"`
`CLI : 4.18.0`
### Schema
```
type Session
  @auth(
    rules: [
      {
        allow: owner
      }
    ]
  )

  @model {
  id: ID!
  owner: String
  title: String
}
```
### Error given 
From AWS AppSync RealTime (background operation for DataStore sync)
`Validation error of type MissingFieldArgument: Missing field argument owner @ 'onCreateSession'"`

I tried many things, custom resolvers... removing the mandatory owner parameter from subscription... but it seems like the other previous issues, that DataStore is not sending the owner field in the subscription

The subscriptions sent by AppSync for DataStore being the following : 
```
"subscription operation {
  onCreateSession {
    id
    owner
    title
    _version
    _lastChangedAt
    _deleted
  }
}
"
``` 
I've read on the Amplify Graph API doc ([here](https://docs.amplify.aws/cli/graphql-transformer/directives#authorizing-subscriptions)) that 
> You don’t need to pass the user as an argument in the subscription request, since the resolver will instead check the contents of your JWT token.
So I also tried manually adding the access token in the header
```
Amplify.configure(awsconfig);
Amplify.configure({
    graphql_headers: async () => {
      const session = await Auth.currentSession();
      return {
        Authorization: session.getAccessToken().getJwtToken()
      };
    }
  }
});
```
But the error is still there, so I don't really know what to do... Did I miss something somewhere (I digged into all the related issues here, and tried all the workaround without success)

**Comment by richardfan1126** [comment]
@tomheno The subscription should receive owner as a parameter, something like:
```
subscription operation($owner: String!) {
  onCreateSession(owner: $owner) {
    id
    owner
    title
    _version
    _lastChangedAt
    _deleted
  }
}
```

How did you generate those codes?

**Comment by tomheno** [comment]
@richardfan1126 Got it from the console, Amplify Logger is logging at DEBUG level, see the following screenshot : 
![image](https://user-images.githubusercontent.com/30948041/80228707-aa8b0f80-864f-11ea-847d-4d3c78bf161b.png)
There's also the same error with onUpdateSession & onDeleteSession subscriptions, that are run by DataStore in background, those messages occures on app start.

**Comment by veproza** [comment]
I ran into pretty much the same issue with `Sync error subscription failed Connection failed: {"errors":[{"message":"Validation error of type MissingFieldArgument: Missing field argument owner @ 'onCreateXXX'`. In the end, I managed to work around it by explicitly specifying the owner field in the schema:

```
type SecondTestField
  @auth(
    rules: [
      {
        allow: owner,
        ownerField: "owner",
        operations: [create, update, delete, read]
      }
    ]
  )
  @model {
  id: ID!
  title: String!
  description: String
  owner: String
}
```

Also during the debugging I realized I forgot to run `amplify codegen models`, so that may also be one of the steps leading to this issue.

But I don't think it's the only one. I now have for testing two models, one with explicit `owner` and one with implicit `owner` and when using the implicit/automatic one, I don't get any errors, but I also don't get subscription functionality. I need to refresh the page each time for new data to load. The implicit model is defined like this:

```
type TestField
  @auth(
      rules: [
        {
          allow: owner
        }
      ]
  )
  @model {
    id: ID!
    title: String!
    description: String
}
```
so the only difference is missing explicit `owner` field. Yet in code, `TestField` doesn't support subscriptions (needs a page reload to display new data) and `SecondTestField` works as it should.

AWS libs are as follows:
```
        "@aws-amplify/api": "^3.1.9",
        "@aws-amplify/datastore": "^2.0.10",
        "@aws-amplify/pubsub": "^3.0.10",
        "@aws-amplify/ui-react": "^0.2.5",
        "aws-amplify": "^3.0.10",
```

**Comment by johanbuys** [comment]
Hi I am having similar issues. 

One @auth rule works:
```
type Task @model 
  @auth(rules: [
      { allow: owner, ownerField: "owner" }
    ]) {
  id: ID!
  title: String!
  owner: String
}
```

But when I want to add a second auth rule
```
type Task @model 
  @auth(rules: [
      { allow: owner, ownerField: "owner" },
      { allow: owner, ownerField: "assignee", operations: [update, read] }
    ]) {
  id: ID!
  title: String!
  owner: String
  assignee: String
}
```

with the following:
`[WARN] 40:44.603 DataStore - Sync error subscription failed Connection failed: {"errors":[{"message":"Validation error of type MissingFieldArgument: Missing field argument owner @ 'onUpdateTask'"}]}`

And does not sync to cloud, local changes are working though. 

```
"dependencies": {
        "@aws-amplify/core": "^3.2.6",
        "@aws-amplify/datastore": "^2.0.10",
        "@aws-amplify/ui-react": "^0.2.5",
```

**Comment by MrHertal** [comment]
I have the same issue.

Using

```
"@aws-amplify/api": "^3.1.10",
"@aws-amplify/auth": "^3.2.7",
"@aws-amplify/core": "^3.2.7",
"@aws-amplify/datastore": "^2.1.0",
```

I have tried the various workarounds found here but it still does not work.

@chuckatc mentioned this post [here](https://medium.com/open-graphql/create-a-multiuser-graphql-crud-l-app-in-5-minutes-with-the-amplify-datastore-902764f27404) where it seems to work, but I noticed that the author is using `npx amplify-app` to init his project, so he does not add an API with `amplify add api`.

I am in the case where I had an API before setting up DataStore, so I used the cli to update my API to use DataStore.

@veproza @tomheno In which case are you? Did you init your project with `npx amplify-app` or did you update your existing API tu use DataStore?

**Comment by lenarmazitov** [comment]
I have similar error. 
Right after authenticate user, I'm trying to perform DataStore.query, and get undefined.
But after waiting a few seconds and reload applications query successfully returns data.

**Comment by yashprit** [comment]
I have similar issues, all models are marked with `@auth`

1. Cognito is setup separately and configured in `aws-exports`
2. AppSync api pushed and configured in `aws-exports`

Most of our configuration are manual.

Tried `amplify api update`.

Also tried DataStore.configure(), which give other error `DataStore - Sync error subscription failed Subscribe only available for AWS AppSync endpoint`.

**Comment by MrHertal** [comment]
> I have the same issue.
> 
> Using
> 
> ```
> "@aws-amplify/api": "^3.1.10",
> "@aws-amplify/auth": "^3.2.7",
> "@aws-amplify/core": "^3.2.7",
> "@aws-amplify/datastore": "^2.1.0",
> ```
> 
> I have tried the various workarounds found here but it still does not work.
> 
> @chuckatc mentioned this post [here](https://medium.com/open-graphql/create-a-multiuser-graphql-crud-l-app-in-5-minutes-with-the-amplify-datastore-902764f27404) where it seems to work, but I noticed that the author is using `npx amplify-app` to init his project, so he does not add an API with `amplify add api`.
> 
> I am in the case where I had an API before setting up DataStore, so I used the cli to update my API to use DataStore.
> 
> @veproza @tomheno In which case are you? Did you init your project with `npx amplify-app` or did you update your existing API tu use DataStore?

I made the test and it has nothing to do with whether you started the project using `npx amplify-app` or you updated an existing API.

I still have this issue in my React Native app, I created a new issue as this one is closed.

https://github.com/aws-amplify/amplify-js/issues/5898

**Comment by rpostulart** [comment]
I have a similar issue with 

`type Quiz
  @model
  @auth(
    rules: [
      { allow: public, provider: iam, operations: [read] }
      { allow: owner }
    ]
  ) {
  id: ID!
  title: String!
  seconds: Int!
  currentQuestion: String
  questionOrder: String
  started: Boolean
  questionTime: Int
  view: Int
  owner: String!
}`

[WARN] 34:16.791 DataStore - Sync error – "Connection failed: {\"errors\":[{\"message\":\"Validation error of type MissingFieldArgument: Missing field argument owner @ 'onUpdateQuiz'\"}]}"

` "@aws-amplify/analytics": "^3.1.15",
    "@aws-amplify/api": "^3.1.15",
    "@aws-amplify/auth": "^3.2.12",
    "@aws-amplify/core": "^3.3.2",
    "@aws-amplify/datastore": "^2.2.2",
    "@aws-amplify/interactions": "^3.1.15",
    "@aws-amplify/predictions": "^3.1.15",
    "@aws-amplify/storage": "^3.2.5",
    "@aws-amplify/ui": "^2.0.2",
    "@aws-amplify/ui-react": "^0.2.8",
    "@aws-amplify/xr": "^2.1.15",
    "@testing-library/jest-dom": "^4.2.4",
    "@testing-library/react": "^9.3.2",
    "@testing-library/user-event": "^7.1.2",
    "array-move": "^2.2.1",
    "aws-amplify": "^3.0.13",
    "aws-amplify-react": "^4.1.12",`

**Comment by alexandprivate** [comment]
Same issue here 
This don't work
`@auth(
        rules: [
            { allow: owner, ownerField: "owner" }
            { allow: owner, ownerField: "tecs", operations: [read, update] }
        ]
    ) {...`
But only one rule works perfect
`@auth(
        rules: [
            { allow: owner, ownerField: "owner" }
        ]
    ) {`

**Comment by lenarmazitov** [comment]
Hello @alexandprivate is there missing comma in auth array? 
Please check with comma, if it works let us know.

**Comment by alexandprivate** [comment]
> Hello @alexandprivate is there missing comma in auth array? 
> Please check with comma, if it works let us know.

Hi there, well yes and not, prettier actually removed  the comma when I save the doc, so I don't that's the issue

**Comment by alexandprivate** [comment]
<img width="618" alt="Screen Shot 2020-06-14 at 19 48 52" src="https://user-images.githubusercontent.com/7695887/84608448-1f244280-ae78-11ea-9be9-2c01e83bc494.png">
Ok guys lets see if we can solve this

**Comment by alexandprivate** [comment]
> "@aws-amplify/core": "^3.2.6",
>         "@aws-amplify/datastore": "^2.0.10",
>         "@aws-amplify/ui-react": "^0.2.5",

Any updates on this?

**Comment by alexandprivate** [comment]
https://github.com/aws-amplify/amplify-cli/issues/3048#issuecomment-624720926 Any updates on this?

**Comment by alexandprivate** [comment]
> Hi I am having similar issues.
> 
> One @auth rule works:
> 
> ```
> type Task @model 
>   @auth(rules: [
>       { allow: owner, ownerField: "owner" }
>     ]) {
>   id: ID!
>   title: String!
>   owner: String
> }
> ```
> 
> But when I want to add a second auth rule
> 
> ```
> type Task @model 
>   @auth(rules: [
>       { allow: owner, ownerField: "owner" },
>       { allow: owner, ownerField: "assignee", operations: [update, read] }
>     ]) {
>   id: ID!
>   title: String!
>   owner: String
>   assignee: String
> }
> ```
> 
> with the following:
> `[WARN] 40:44.603 DataStore - Sync error subscription failed Connection failed: {"errors":[{"message":"Validation error of type MissingFieldArgument: Missing field argument owner @ 'onUpdateTask'"}]}`
> 
> And does not sync to cloud, local changes are working though.
> 
> ```
> "dependencies": {
>         "@aws-amplify/core": "^3.2.6",
>         "@aws-amplify/datastore": "^2.0.10",
>         "@aws-amplify/ui-react": "^0.2.5",
> ```

Hi there @johanbuys any updates on this from your side?

**Comment by johanbuys** [comment]
> > Hi I am having similar issues.
> > One @auth rule works:
> > ```
> > type Task @model 
> >   @auth(rules: [
> >       { allow: owner, ownerField: "owner" }
> >     ]) {
> >   id: ID!
> >   title: String!
> >   owner: String
> > }
> > ```
> > 
> > 
> > But when I want to add a second auth rule
> > ```
> > type Task @model 
> >   @auth(rules: [
> >       { allow: owner, ownerField: "owner" },
> >       { allow: owner, ownerField: "assignee", operations: [update, read] }
> >     ]) {
> >   id: ID!
> >   title: String!
> >   owner: String
> >   assignee: String
> > }
> > ```
> > 
> > 
> > with the following:
> > `[WARN] 40:44.603 DataStore - Sync error subscription failed Connection failed: {"errors":[{"message":"Validation error of type MissingFieldArgument: Missing field argument owner @ 'onUpdateTask'"}]}`
> > And does not sync to cloud, local changes are working though.
> > ```
> > "dependencies": {
> >         "@aws-amplify/core": "^3.2.6",
> >         "@aws-amplify/datastore": "^2.0.10",
> >         "@aws-amplify/ui-react": "^0.2.5",
> > ```
> 
> Hi there @johanbuys any updates on this from your side?

Hi, no not really.

I still have the issue. It looks like, If I change around the 2 rules the error changes to match the first field. 

This was core to our use case, I did not want to hack something together to make this work so I haven't looked into this for a while. I am monitoring this thread, but I had to refactor my code, not using Datastore for now.

Hopefully these issues can be fixed I was really eager to use this, but for now it's a showstopper.

**Comment by alexandprivate** [comment]
> > > Hi I am having similar issues.
> > > One @auth rule works:
> > > ```
> > > type Task @model 
> > >   @auth(rules: [
> > >       { allow: owner, ownerField: "owner" }
> > >     ]) {
> > >   id: ID!
> > >   title: String!
> > >   owner: String
> > > }
> > > ```
> > > 
> > > 
> > > But when I want to add a second auth rule
> > > ```
> > > type Task @model 
> > >   @auth(rules: [
> > >       { allow: owner, ownerField: "owner" },
> > >       { allow: owner, ownerField: "assignee", operations: [update, read] }
> > >     ]) {
> > >   id: ID!
> > >   title: String!
> > >   owner: String
> > >   assignee: String
> > > }
> > > ```
> > > 
> > > 
> > > with the following:
> > > `[WARN] 40:44.603 DataStore - Sync error subscription failed Connection failed: {"errors":[{"message":"Validation error of type MissingFieldArgument: Missing field argument owner @ 'onUpdateTask'"}]}`
> > > And does not sync to cloud, local changes are working though.
> > > ```
> > > "dependencies": {
> > >         "@aws-amplify/core": "^3.2.6",
> > >         "@aws-amplify/datastore": "^2.0.10",
> > >         "@aws-amplify/ui-react": "^0.2.5",
> > > ```
> > 
> > 
> > Hi there @johanbuys any updates on this from your side?
> 
> Hi, no not really.
> 
> I still have the issue. It looks like, If I change around the 2 rules the error changes to match the first field.
> 
> This was core to our use case, I did not want to hack something together to make this work so I haven't looked into this for a while. I am monitoring this thread, but I had to refactor my code, not using Datastore for now.
> 
> Hopefully these issues can be fixed I was really eager to use this, but for now it's a showstopper.

Oh! Sad to hear it, I would like to keep DataStore but implicates to hacking the request buy user and basically remove the @auth from the affected model, let's hope this could be solved sooner than expected.

@elorzafe any updates on this so far?

**Comment by GavinSawyer** [comment]
Hi everybody, I've been having this issue. I'm noticing that it basically only happens when you limit certain operations. Having only operations: ["read"] for me, was producing errors for "onCreate{type}", "onUpdate{type}", and on "onDelete{type}". It seems that the Amplify CLI is generating more than it needs to in subscriptions.graphql and maybe elsewhere. I haven't looked into this more as I just found this out, but it's worth trying if you think it helps! I'll comment again if I can figure out more!

**Comment by johanbuys** [comment]
> Hi everybody, I've been having this issue. I'm noticing that it basically only happens when you limit certain operations. Having only operations: ["read"] for me, was producing errors for "onCreate{model}", "onUpdate{model}", and on "onDelete{model}". When I added "update" and "delete" operations, only the error for "onCreate{model}" appeared.
> 
> It seems that the Amplify CLI is generating more than it needs to in subscriptions.graphql and maybe elsewhere. I haven't looked into this more as I just found this out, but it's worth trying if you think it helps! I'll comment again if I can figure out more.
> 
> edit: In case you're wondering, my model was set to only have operations: ["read"] overall despite my project needing to track changes in real time. A few of the items also had "update", but apparently the object overall needs to adopt all operations for DataStore to work.

This is also what I have found. See this https://docs.amplify.aws/cli/graphql-transformer/directives#multiple-authorization-rules

```
type Draft @model
  @auth(rules: [
    # Defaults to use the "owner" field.
    { allow: owner },

    # Authorize the update mutation and both queries.
    { allow: owner, ownerField: "editors", operations: [update, read] }
  ]) {
  id: ID!
  title: String!
  content: String
  owner: String
  editors: [String]
}
```
And this translates to: 

> Draft that stores unfinished posts for a blog. You might want to allow the Draft’s owner to create, update, delete, and read Draft objects. However, you might also want the Draft’s editors to be able to update and read Draft objects.

This is the exact use case I was trying to accomplish and it did not work.

**Comment by rpostulart** [comment]
@yuth @elorzafe @undefobj 
Can you re-open this issue? It is still not solved in the last relase!

**Comment by undefobj** [comment]
Reopening cc @sammartinez @manueliglesias @iartemiev

**Comment by sammartinez** [comment]
Thanks @undefobj, Im going to transfer this to the Amplify JS Repo so we can dig into it
